### ISSUE TITLE
tensorflow 2.0 strategy for tensorflow-io

### ISSUE THREAD

**Issue by yongtang** [tensorflow/io#411]
There are lots of changes in tensorflow for 2.0 recently and that effectively makes tensorflow-io build break against tf-nightly and tf-nightly-2.0-preview (see #410 for an example).

As 2.0 release is near (see https://github.com/tensorflow/io/issues/403), we have to come up with a 2.0 strategy that is feasible.

Here are the list of issue we encounter, once we bump the version from tensorflow==2.0.0beta1 to tensorflow==nightly:

1) protobuf has to be updated to 3.8.0 for TF 2.0 and stay 3.7.1 for TF 1.14.x.
2) grpc has to be updated to 1.22.0+ for TF 2.0 and stay with 1.21.1 for TF 1.14.x
3) tf.data.Dataset's base class now requires `element_spec` for TF 2.0 and stays with `element_strucutre` for 1.14.x
4) we also have to drop `output_types`, `output_shapes` pair and replace them with `TensorSpec`.
5) tf.data.Dataset's input method may need some adjustment in TF 2.0, haven't figure out exact change yet.

@terrytangyuan @BryanCutler @dmitrievanthony I am opening up this issue, as it looks like we will be forced to either stay with 1.14 or go completely to TF 2.0 very soon (in a couple of weeks).

My thinking at the moment, is that we release another version tensorflow-io 0.8.0 to stay with 1.14.x.

Then next week after tensorflow's nightly switching to 2.0 by default (https://github.com/tensorflow/io/issues/403#issue-475662297), we will switch to 2.0 by default as well, fix the above mentioned issues, and drop 1.14.x support for tensorflow-io. WDYT?

**Comment by yongtang** [comment]
I am also thinking we should stick with primitive tensor ops as the upstream API is quite stable than dataset. @BryanCutler @terrytangyuan

**Comment by terrytangyuan** [comment]
+1 to switch to 2.0 by default and use of primitive tensor ops.

**Comment by yongtang** [comment]
@BryanCutler @terrytangyuan Release candidate of TF 2.0.0 is imminent:
https://groups.google.com/a/tensorflow.org/forum/#!topic/developers/SIfaft4wuXY

We will need to get manylinux2010 working soon.

I have been trying to build manylinux2010. I believe it is very close to get it done.

**Comment by BryanCutler** [comment]
I think having a final tf-io 0.8.0 for 1.14.x would be good. After that, I agree we should switch gears to concentrate on TF 2.0.

**Comment by yongtang** [comment]
Update on manylinux2010 and TF 2.0:

Essentially we are able to generate manylinux2010 and it works fine (PR #410):
```
Fixed-up wheel written to /v/wheelhouse/tensorflow_io-0.9.0-cp37-cp37m-manylinux2010_x86_64.whl

ls wheelhouse/*

wheelhouse/tensorflow_io-0.9.0-cp27-cp27mu-manylinux2010_x86_64.whl
wheelhouse/tensorflow_io-0.9.0-cp35-cp35m-manylinux2010_x86_64.whl
wheelhouse/tensorflow_io-0.9.0-cp36-cp36m-manylinux2010_x86_64.whl
wheelhouse/tensorflow_io-0.9.0-cp37-cp37m-manylinux2010_x86_64.whl
```

The only issue is that there are quite some API changes in TF 2.0. The plan
is that we will wait until a release candidate of TF 2.0 out to make another push to conform to manylinux2010.

**Comment by yongtang** [comment]
The information is a little confusing:
https://groups.google.com/a/tensorflow.org/forum/#!topic/developers/tds7vLcmIrA

as tf-nightly is still updating after 08/16.

But, my guess is that TF 1.15 will be manylinux2010.

Base on that, and it is harder and harder to support
- manylinux2010(1.15)
- manylinux1 (1.14)
- 2.0.0Beta1 (2.0+manylinux1)
- 2.0.0RC (2.0+manylinux2010), 

Plus recent API changes in tf.data (element_strucutre => element_spec), here is the plan:
1)  On master, switch to tf-nightly until 1.15RC is out
2) After 1.15RC is out, cut a branch of R0.8 for 1.x support
3) On master, switch to tf-nightly-2.0-preview, and switch to 2.0RC when it is out.
4) Drop 1.x components in master
5) Release a final 0.8.0 version, and merge R0.8 branch back to master.
6) Release a 0.9 version that support 2.0 only

**Comment by yongtang** [comment]
@terrytangyuan @BryanCutler We are almost there with manylinux2010. In PR #426, we get the binary packed into manylinux2010, though there is one limitation.

Since tensorflow whl package carries libtensorflow.so file, and out whl have to depends on this  libtensorflow.so file, when we try to use `auditwheel` to repair our whl for manylinxu2010, it will copy the libtensorflow.so file. That means in a system, user will have libtensorflow.so in tensorflow package and another copy of libtensorflow.so in tensorflow-io package.

This is not going to work well.

So while tensorflow's whl has been updated to manylinux2010, our package of whl (tensorflow-io) could not conform to manylinux2010.

I am pretty sure the same issue will apply to any other whl packages in the whole ecosystem that depends on tensorflow's so file....

The tensorflow-io's whl could get back to manylinux1 which does not have this limitation....

I will try to do some reverse engineering of auditwheel to see if I could find a workaround with some hack.

**Comment by yongtang** [comment]
TensorFlow 2.0.0RC0 has been released. I created a Working in Progress PR #444. We still have to wait for TensorFlow 1.15.0RC0 to merge. Though.

**Comment by yongtang** [comment]
@BryanCutler  @terrytangyuan 

See #444 for related discussions.

At the moment:
1) master branch is based on tf-nightly and we build against tf-nightly only
2) tf-nightly stopped updating in PyPI.org since 08/21
3) tensorflow==2.0.0rc0 has been released
4) tensorflow==1.15.0 (or RC) has not been released yet.
5) We are waiting for tensorflow==1.15.0 release to build tensorflow-io 0.8.0

However, there is a growing need (See https://github.com/tensorflow/io/pull/444#issuecomment-528900723) to have tensorflow==2.0.0 as default (as opposed to 1.15.0). We also need to cover yet to be released 1.15.0.  I think the following plan will work:

1) We switch to tensorflow==2.0.0rc0 in master branch and for all nightly build
2) We create a R0.8 branch that ties to tf-nightly (and wait for 1.15.0 release)
3) The created R0.8 branch will have no change except for version change
4) A PR will be created from R0.8 and will be rebased often so that it will be constantly tested with Travis CI and Kokoro CI.

@terrytangyuan @BryanCutler I think the above plan will work and serve different version requirement (until 1.15.0 is released).

**Comment by BryanCutler** [comment]
@yongtang that sounds like a good plan to me.  Just one minor suggestion, could we call the branch something general like `branch-0.8` and have that serve as a maintenance branch in case we need to do bug fixes and might want to make a 0.8.1 release?

**Comment by terrytangyuan** [comment]
Agreed. We should probably start creating release branches so we could merge in hotfixes later if needed as well as making them as protected branches.

**Comment by yongtang** [comment]
@terrytangyuan @BryanCutler Thanks. Currently the name of the branch (for 0.8) is named as `R0.8` which matches the naming in tensorflow core repo. We can change the name to `branch-0.8` if we prefer.

Also, it looks like `1.15.0rc0` has been released so I bumped the `R0.8` to use `tensorflow==1.15.0rc0`.

**Comment by BryanCutler** [comment]
Oh sure, the name really isn't important as long as we can use it for fixes and maintenance release later, if needed. Thanks @yongtang !

**Comment by yongtang** [comment]
With tensorflow-io 0.9.0 released (support tensorflow 2.0), this issue has been resolved.
